HTTP facility supports UNIX sockets

daemianmack opened the pull request:
This PR modifies the CLJS HTTP interface and the underlying C wrapper around libcurl to add support for [UNIX sockets](https://curl.haxx.se/libcurl/c/CURLOPT_UNIX_SOCKET_PATH.html).

The use case prompting this change: talking to a socket via Planck lets us easily automate Docker using its [HTTP API](https://docs.docker.com/develop/sdk/examples/) -- particularly valuable when operating from within AWS ECS to avoid "Docker-in-Docker" woes.

I _think_ the C code here is correct, but it's been approximately forever since I've written any, so extra suspicion is appreciated. :)

mfikes commented:
@daemianmack I see no fundamental problem with that. Looking at the Travis build failure, I wonder if this relies on a newer capability in libcurl that doesn't exist on the Travis build boxes.

The bit of build code that specifies that Planck depends on curl is here
https://github.com/mfikes/planck/blob/master/planck-c/CMakeLists.txt#L58

Perhaps there might be some way to conditionally opt out of the feature if it is not present in whatever older libcurl is on the box where Planck is being built? (Just thinking out loud.)

daemianmack commented:
Hey, I pushed a fix before I saw you had closed this. Not sure if this is obviated by FindCURL, but libcurl has a decent [mechanism for determining feature presence](https://curl.haxx.se/libcurl/c/curl_version_info.html) which I've made use of in that commit.

If you'd prefer to go the FindCURL route, the version boundary important to UNIX socket support is [7.4.0](https://curl.haxx.se/changes.html)...

```
Fixed in 7.40.0 - January 8 2015
Changes:
...
Added support for HTTP over unix domain sockets, via CURLOPT_UNIX_SOCKET_PATH and --unix-socket
```

mfikes commented:
@daemianmack Nah, your change is good. I actually didn't close this issue—it was the "Use FindCURL" ticket which was closed.

mfikes commented:
@daemianmack I've long forgotten what is the best way to do things in C. :) I don't at all mind things being awkward in C so long as things work. (Planck is a smallish codebase anyways.)

mfikes commented:
@daemianmack My plan is to add a CHANGELOG.md entry as well as the associated revisions to the site docs here https://github.com/mfikes/planck/blob/master/site/src/planck-http.md followed by releasing an update to Planck when the next ClojureScript release is made. Let me know if you'd like a release of Planck prior to then and I can cut one.

daemianmack commented:
That sounds good -- right now I'm the only one needing this enhancement and I'm fine working with a locally-crafted binary. If we hit the rollout date prior to the next CLJS release, I'll ping you to see about a Planck release.

In the meantime, I'm puzzled by these CI failures -- the two Linux ones passed, while the two OSX ones failed. I've been progressively testing this patch on both OSX (v 10.11.6) and Linux (an unreasonably ancient Ubuntu) and haven't encountered segfaults that weren't the result of forgetting to do a `script/clean` prior to a `script/build && script/test`.

I'm going to try and figure out what's different between my OSX dev environment and the environment that Travis is setting up. If you have any pointers here, I'd appreciate them!

mfikes commented:
I am able to get it to crash on macOS 10.13.1 when running the tests, and started debugging it a little. I found that if I turn off optimizations (set Debug instead of Release here https://github.com/mfikes/planck/blob/master/planck-c/CMakeLists.txt#L4) it goes away. The function that it crashes in already has optimizations selectively disabled via this bit of code https://github.com/mfikes/planck/blob/master/planck-c/http.c#L83-L91 so I started down the path of adding `fprintf(stderr, "here")` style statements to try to figure out precisely where it is crashing when in Release mode without the benefit of really being able to use a debugger. I won't be able to get back to this for a bit, but will resume in a day or so if you haven't sorted it by then.

daemianmack commented:
I was unable to reproduce on OSX 10.11.6.

I *was* able to reproduce in a Virtualbox install of OSX 10.12.6, with the same behavior you report regarding "Debug" vs "Release".

It looks like the segfault occurs the [first time](https://github.com/daemianmack/planck/blob/f87edb071cbd46b5f1c52b9a5dc5d1ee6afcc424/planck-c/http.c#L210) we pass `result` to `JSObjectSetProperty`. 

That line happens to reference `body` but I moved the corresponding `JSObjectSetProperty` for `status` above it (i.e., bubbled [L215](https://github.com/daemianmack/planck/blob/f87edb071cbd46b5f1c52b9a5dc5d1ee6afcc424/planck-c/http.c#L215) to L210) and then the segfault occurred there instead.

mfikes commented:
@daemianmack Earlier I had seen some extremely odd segfaulting behavior that I ultimately chalked up as an optimizer bug. (The behavior involved a segfault going away by simply re-arranging some code—copying fields of a struct to some locals and holding onto them earlier as opposed to later.) These shenanigans have only occurred in `http.c`, to my memory, and I ultimately took the route of inhibiting optimization in the places that failed. In truth, the problem could be rooted in some code that is arguably incorrect, and only works because it is lucky when un-optmized, but to date, I haven't succeeded in finding anything truly wrong with the code.

What seems to be odd about this new segfault is that the failure is occurring in code that is already marked to not be optimized. Hopefully I'm wrong (which could easily be the case given how little I've looked into it), and the solution might be simply to increase the scope of what is not being optimized. If we ultimately end up marking the entire `http.c` file as something that should not be optimized, I think that would probably be OK, given that the file is really about dealing with I/O anyway.

daemianmack commented:
I can confirm that the segfault goes away when I temporarily re-instate [this change](https://github.com/mfikes/planck/commit/217a85f21cd7123aa20c716ed513f4b5d62bc3ee). As I understand, that change was reverted because it's not respected by Homebrew, and thus only benefits those manually compiling for themselves. Do you have a clear idea for how we should proceed?

mfikes commented:
@daemianmack If you don't mind, I'll take on the responsibility of figuring out what is going wrong here. Your code revision is essentially a simple enhancement and I feel like you've stepped into a problem that speaks to the fragility of the band-aids that have been put on in the past.

mfikes commented:
This issue is essentially blocked by #600. I'll see if I can sort out what is going on via that ticket.

daemianmack commented:
Cool -- I'm shooting in the dark at the moment. Rope me in if there's anything I can do to help!

daemianmack commented:
Awesome -- thanks!

daemianmack commented:
I'm at the point where I could use a release, if convenient for you -- I used the local ubuntu_16 Vagrant build but it's requiring a bunch of lib paths (specifically, those for `libzip` and `libicu`) symlinked on the target machine.

mfikes commented:
@daemianmack The Planck 1.11.0 PPA is now available for Ubuntu 16.04.

daemianmack commented:
Thanks very much!
